contracts/adapters/IPRebalanceExtension.sol

```diff
@@ -0,0 +1,478 @@
+
+                uint256 targetUnderlying = rebalanceParams[component];
+                int256 diff = finalTotalUnderlyingUnits.toInt256() - currentTotalUnderlyingUnits.toInt256() + targetUnderlying.toInt256();
+
+                rebalanceTargets[i] = diff > 0 ? diff.toUint256() : 0;
+            }
+        }
+
+        (
+            address[] memory newComponents,
+            uint256[] memory newComponentsTargetUnits,
+            uint256[] memory oldComponentsTargetUnits
+        ) = _sortNewAndOldComponents(setComponentList, rebalanceTargets);
+
+        bytes memory callData = abi.encodeWithSelector(
+            IGeneralIndexModule.startRebalance.selector,
+            setToken,
+            newComponents,
+            newComponentsTargetUnits,
+            oldComponentsTargetUnits,
+            setToken.positionMultiplier()
+        );
+
+        invokeManager(address(generalIndexModule), callData);
+    }
+
+    /**
+     * Absorbs all valid airdrops passed into the components parameter. If a component passed in is not a
+     * valid airdrop component, ignore it.
+     */
+    function _absorbAirdrops(address[] memory _components) internal {
+        address[] memory airdropTokens = airdropModule.getAirdrops(setToken);
+        address[] memory tokensToAbsorb = new address[](_components.length);
+
+        uint256 numTokensToAbsorb = 0;
+        // can do this is n*log(n) time using a sort but might not be worth the effort
+        for (uint256 i = 0; i < _components.length; i++) {
+            if (airdropTokens.contains(_components[i])) {
+                tokensToAbsorb[numTokensToAbsorb] = _components[i];
+                numTokensToAbsorb++;
+            }
+        }
+
+        if (numTokensToAbsorb == 0) return;
+
+        address[] memory batchAbsorbTokens = new address[](numTokensToAbsorb);
+        for (uint256 i = 0; i < numTokensToAbsorb; i++) {
+            batchAbsorbTokens[i] = tokensToAbsorb[i];
+        }
+
+        bytes memory callData = abi.encodeWithSelector(airdropModule.batchAbsorb.selector, setToken, batchAbsorbTokens);
+        invokeManager(address(airdropModule), callData);
+    }
+
+    /**
+     * Checks if a component is a valid transform component. This can be done by checking if it has a TransformInfo
+     * entry.
+     */
+    function _isTransformComponent(address _component) internal view returns (bool) {
+        return transformComponentInfo[_component].underlyingComponent != address(0);
+    }
+
+    /**
+     * Gets the total amount of a underlying component in the target set units. This value includes all units present due to
+     * a transform component containing it as its underlying. Units from the raw underlying component potentially being a component
+     * in the set are ignored.
+     */
+    function _getFinalTotalUnderlyingUnits(address _underlying, address[] memory _components) internal view returns (uint256) {
```

Comment:
So this exists because we may wrap the same underlying component into multiple different assets?

Reply:
Yep. This is already a requirement for some products in the pipeline, but does unfortunately complicate the accounting a bit